src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -480,13 +478,148 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
+                pingContent.CopyTo(_outgoingBuffer.AvailableMemory);
+                _outgoingBuffer.Commit(FrameHeader.PingLength);
+
+                await FlushOutgoingBytesAsync().ConfigureAwait(false);
+            }
+            finally
+            {
+                ReleaseWriteLock();
+            }
+        }
+
+        private async ValueTask SendRstStreamAsync(int streamId, Http2ProtocolErrorCode errorCode)
+        {
+            await AcquireWriteLockAsync().ConfigureAwait(false);
+            try
+            {
+                WriteFrameHeader(new FrameHeader(FrameHeader.RstStreamLength, FrameType.RstStream, FrameFlags.None, streamId));
+
+                _outgoingBuffer.EnsureAvailableSpace(FrameHeader.RstStreamLength);
+                _outgoingBuffer.AvailableSpan[0] = (byte)(((int)errorCode & 0xFF000000) >> 24);
+                _outgoingBuffer.AvailableSpan[1] = (byte)(((int)errorCode & 0x00FF0000) >> 16);
+                _outgoingBuffer.AvailableSpan[2] = (byte)(((int)errorCode & 0x0000FF00) >> 8);
+                _outgoingBuffer.AvailableSpan[3] = (byte)((int)errorCode & 0x000000FF);
+                _outgoingBuffer.Commit(FrameHeader.RstStreamLength);
+
+                await FlushOutgoingBytesAsync().ConfigureAwait(false);
+            }
+            finally
+            {
+                ReleaseWriteLock();
+            }
+        }
+
+        private (ReadOnlyMemory<byte> first, ReadOnlyMemory<byte> rest) SplitBufferForFraming(ReadOnlyMemory<byte> buffer) =>
+            buffer.Length > FrameHeader.MaxLength ?
+                (buffer.Slice(0, FrameHeader.MaxLength), buffer.Slice(FrameHeader.MaxLength)) :
+                (buffer, Memory<byte>.Empty);
+
+        private async ValueTask SendStreamDataAsync(int streamId, ReadOnlyMemory<byte> buffer)
+        {
+            ReadOnlyMemory<byte> remaining = buffer;
+
+            while (remaining.Length > 0)
+            {
+                ReadOnlyMemory<byte> current;
+                (current, remaining) = SplitBufferForFraming(remaining);
+
+                await AcquireWriteLockAsync().ConfigureAwait(false);
+                try
+                {
+                    _outgoingBuffer.EnsureAvailableSpace(FrameHeader.Size + current.Length);
+                    WriteFrameHeader(new FrameHeader(current.Length, FrameType.Data, FrameFlags.None, streamId));
+                    current.CopyTo(_outgoingBuffer.AvailableMemory);
+                    _outgoingBuffer.Commit(current.Length);
+
+                    await FlushOutgoingBytesAsync().ConfigureAwait(false);
+                }
+                finally
+                {
+                    ReleaseWriteLock();
+                }
+            }
+        }
+
+        private async ValueTask SendEndStreamAsync(int streamId)
+        {
+            await AcquireWriteLockAsync().ConfigureAwait(false);
+            try
+            {
+                _outgoingBuffer.EnsureAvailableSpace(FrameHeader.Size);
```

Comment:
This `EnsureAvailableSpace` check is redundant, because the `WriteFrameHeader()` in the next line will do it.